### PR TITLE
Add pretrained representation and transfer learning framework

### PR DESCRIPTION
**Pull request by Kai-Zhu-2001**
## Description

This PR introduces a general representation framework for reusing pretrained models in `mlcolvar`.

The new API separates reusable representations from task-specific heads and provides a common interface for both external atomistic models and existing `mlcolvar` models.

Main features include:

- a unified `mlcolvar.representation` API;
- common `Representation`, `VectorRepresentation`, and `GraphRepresentation` interfaces;
- MACE as a reference adapter for external pretrained atomistic models;
- `MLColvarRepresentation` for reusing existing vector- and graph-based `mlcolvar` models;
- composable graph transformations through `.pool()` and `.concat_atoms()`;
- frozen pretrained representations with trainable task-specific heads;
- representation and Jacobian caching for efficient downstream training;
- public `precompute_committor_cache` support for cached committor training while preserving derivatives with respect to the original inputs;
- preserved descriptor and coordinate gradients through frozen representations;
- support for periodic systems, variable-size graph batches, mixed precision, and element alignment;
- TorchScript and Metatomic export for deployment.

External pretrained atomistic models can be integrated by subclassing the common `GraphRepresentation` interface, with MACE providing a concrete reference implementation.

`MACERepresentation` targets the standard `mace-torch` model interface. Compatible models are expected to expose their atomic species, cutoff, interaction layers, and `node_feats` output. Automatic descriptor-layout inference additionally relies on the standard MACE `products[0].linear.irreps_out` structure, while `num_features` and `l_max` can be provided explicitly for custom variants.

## Main changes

- [x] Add the `mlcolvar.representation` module
- [x] Add `Representation`, `VectorRepresentation`, and `GraphRepresentation`
- [x] Add `MACERepresentation` as the reference external atomistic adapter
- [x] Add `MLColvarRepresentation` for existing `mlcolvar` models
- [x] Add composable `.pool()` and `.concat_atoms()` operations
- [x] Add reusable `TaskHead` and `RepresentationModel`
- [x] Add representation and Jacobian caching through `.cache()`
- [x] Add public `precompute_committor_cache` support for cached committor training
- [x] Preserve descriptor and coordinate gradients for frozen representations
- [x] Add TorchScript export utilities
- [x] Add Metatomic export with automatic metadata inference
- [x] Add unit tests
- [x] Add tutorials and documentation

## API design

The framework follows a representation-first workflow:

```text
raw input
    ↓
Representation
    ↓
optional representation transform
(.pool() / .concat_atoms())
    ↓
TaskHead
    ↓
output
```

For example, a pretrained MACE model can be used as:

```python
representation = MACERepresentation(
    pretrained_model,
    freeze=True,
).pool("mean")

head = TaskHead(
    representation.out_features,
    n_out=1,
)

model = RepresentationModel(
    representation,
    head=head,
)
```

Existing `mlcolvar` models use the same representation interface:

```python
representation = MLColvarRepresentation(
    pretrained_cv,
    mode="latent",
    freeze=True,
)

model = RepresentationModel(
    representation,
    n_out=1,
)
```

Frozen representations can be precomputed when repeated evaluation of the pretrained backbone is unnecessary:

```python
cache = representation.cache(
    dataset,
    device="cuda",
)
```

For gradient-based downstream objectives, representation Jacobians can also be cached and reused through the chain rule. For graph representations, Jacobian caching currently requires the selected graphs to contain the same number of atoms because the coordinate Jacobians are stored as a dense tensor. Feature-only graph caching supports variable-size systems.

Cached representations can also be prepared directly for committor training:

```python
cached_dataset, derivatives = precompute_committor_cache(
    representation,
    dataset,
)
```

This allows the same pretrained representation to be reused across different learning objectives while keeping representation extraction, representation transforms, and task-specific heads separate.

## Status

- [x] Implementation
- [x] Tests
- [x] Tutorials
- [x] Documentation
- [x] MACE workflow
- [x] TorchScript export
- [x] Metatomic export
- [x] Ready for review

### CODE REVIEW

**Review by Katomoto — COMMENTED** [review]
Approach is solid. Minor nit: this would benefit from a brief comment explaining why we skip validation for internal callers.

**Comment by Kai-Zhu-2001** [comment]
The tutorial requires optional external dependencies, including MACE and Metatomic, which are not installed in the standard mlcolvar CI environment. It also downloads pretrained external models, making the test slower and dependent on network availability.

Would it be better to exclude this advanced integration tutorial from the standard notebook CI and rely on the dedicated unit tests for the integration layer? We could later add a separate optional-dependencies workflow to test the full tutorial with MACE and Metatomic installed.

@luigibonati

**Review by volkkov — COMMENTED** [review]
Logic looks correct. One suggestion: the early return on line prevents the cleanup function from running — worth adding a finally block.

**Comment by Kai-Zhu-2001** [comment]
If MACE is used both as the PES and as the CV backbone, the model currently needs to be loaded twice, which may significantly increase GPU memory usage. Could the interface eventually support a shared MACE backbone that outputs both the energy and the CV features, allowing the model to be loaded and evaluated only once? I realize that this may require coordination between the MD engine, Metatomic, and PLUMED, but it could be useful to keep this use case in mind.

**Review by volkkov — COMMENTED** [review]
The happy path is clean. What happens if the third-party API returns 429 here? Might be worth handling that case explicitly.

**Review by Deniwn22 — COMMENTED** [review]
Logic looks correct. One suggestion: the early return on line prevents the cleanup function from running — worth adding a finally block.
